### ISSUE TITLE
Implement lua __index and __newindex for capi objects

### ISSUE THREAD

**Issue by Elv13** [awesomeWM/awesome#163]
client, tag, screen, drawable, key and drawin should have their (unknown) property handler delegated to lua.

This API could be used to:
* Implement dynamic properties (replace awful.{...}.getproperty / setproperty)
* Forward property to awful.{client, tag,...}.set_ and get_ when they exist
* Other use

Those handlers should be per object (because someone will eventually complain if they aren't) but always fallback to a common (per class) handler.

New API for all objects:
.index_handler (property with set and get)
.newindex_handler (property with set and get)

**Comment by Elv13** [comment]
```[11:30] <psychon> Elv1313: that would make the code for setting titlebars harder to use / kinda ugly
[11:30] <psychon> I wonder if we can come up with something that is "transparent" to uses in rc.lua
[11:31] <Elv1313> psychon: yes, but it will require being able to set custom metatables for client
[11:31] <Elv1313> that way we can use __index to hijack .border_width and __newindex for the setter. Then it wont break the API and just work
[11:31] <normalra> Exactly.
[11:31] <psychon> hm... might be a good idea to have
[11:32] <psychon> well, rather than setting the metatable, it might be easier to be able to set handlers for __index and __newindex for "property misses" (when the C code doesn't know how to handle things)
[11:32] <Elv1313> I tried to implement that the other day, but ran out of time without much results (I don't know the luac API that well)
[11:32] <Elv1313> psychon: yes, that would work
[11:35] <psychon> "just" requires patching luaA_class_newindex() and luaA_class_index()
[11:35] <psychon> instead of calling newindex_miss_property, they'd need to call into some lua function (that somehow lua must be able to set earlier...)
[11:38] <Elv1313> psychon: if you take care of the C side, I will implement the handlers. I already have the code in Radical to do just that
[11:38] <Elv1313> it also emulate :set_foo(x) -> object.foo = x
[11:41] <Elv1313> psychon: do you want me to create an issue for that?
[11:41] <psychon> yeah, feel free
[11:41] <psychon> and we need to come up with some API for lua to set the handlers
[11:41] <psychon> does it make sense to have them object-specific? or is it ok to have it just per-class?
[11:42] <blueyed> psychon: good reasons against such a tag.  It would be nice for "git describe", e.g. for the awesome-git version in Nix/NixOS, but I can use a custom version like "3.6-git-g4636b11" myself.
[11:43] <Elv1313> I guess client, tag, screen, drawable, key and drawin should have the same API
[11:43] <psychon> looking forward to your feature requests which includes explanations on how the API might look like :-P
[11:47] <phryk> Do things like wibox or the widgets use lua classes and I just find lua confusing or are they implemented as just a bunch of functions?
[11:48] <Elv1313> psychon: https://github.com/awesomeWM/awesome/issues/163
[11:48] <psychon> phryk: lua doesn't have a built-in concept of a "class"
[11:48] <psychon> Elv1313: thanks, will take a look on saturday

```

**Comment by psychon** [comment]
See https://github.com/psychon/awesome/commit/e0913818aa39d8c30cdead8668b5193e26f3d699 and the following code that can be added to the default config:

    -- For experiments with the error handling:
    local check_errors = false
    
    if check_errors then
        drawin.set_index_miss_handler(function(...) error("Nope") end)
        drawin.set_newindex_miss_handler(function(...) error("Nope") end)
    else
        drawin.set_index_miss_handler(function(...) print("index miss", ...) return print, 42 end)
        drawin.set_newindex_miss_handler(function(...) print("newindex miss", ...) return 42 end)
    end
    
    d = drawin({})
    print("getting d.foo")
    print(d.foo)
    print("setting d.bar = 7")
    d.bar = 7
    -- This would call nil
    if not check_errors then
        print("calling foobar")
        print(d:foobar("calling foobar"))
    end

I used `.set_index_miss_handler(func)` instead of `.index_miss_handler = func` because this is easier on the C side. There is already support for adding functions to all objects (for the `.instances()` and signal handling functions), but hijacking `_index` and `_newindex` on classes would be more complicated.

I didn't add a getter because I'm lazy and would rather have a single place in lua where something is implemented based on this instead of modules "chaining calls" to each other (get old handler, set your own which calls the old handler if it doesn't know what to do).

I didn't add this `set_index_miss_handler` to individual objects, but only to classes (`c` vs `client`), since I'm lazy and this approach here is already complicated enough IMO.

Does this still seem ok? Anyone wants to experiment with this code?

**Comment by psychon** [comment]
See the discussion at https://github.com/awesomeWM/awesome/pull/482 and especially https://github.com/awesomeWM/awesome/pull/482#issuecomment-170759662.

**Comment by psychon** [comment]
Fixed with #482 and #757
